### PR TITLE
Feature insertreplace partition

### PR DESCRIPTION
**Pull request by dreamquster**
### What problem does this PR solve? 
Implement insert and replace for hash partitioned table:
https://github.com/pingcap/tidb/issues/8377
### What is changed and how it works?
Add partition check before insert or replace statement execution

### Check List 

Tests 

 - Unit test
  insert_test.go-TestInsertWithPartition

Code changes
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Possible performance regression

Related changes
 - Need to update the documentation
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8569)

### CODE REVIEW

**Comment by jackysp** [comment]
Thanks @dreamquster !
Please fix CI.

File: table/table.go [hunk 1]
```diff
@@ -154,6 +160,8 @@ type Table interface {
 
 	// Type returns the type of table
 	Type() Type
+
+	LocationPartition(ctx sessionctx.Context, r []types.Datum) (int, error)
```
Comment (tiancaiamao) [inline table/table.go]: `LocationPartition` should not be put in the `Table` interface.
There is a `PartitionedTable` for it.

File: executor/insert_common.go [hunk 2]
```diff
@@ -579,3 +588,36 @@ func (e *InsertValues) addRecord(row []types.Datum) (int64, error) {
...
+	}
+
+	for _, insertPartitionName := range e.PartitionNames {
+		if _, ok := tablePartitionNames[insertPartitionName]; !ok {
+			return table.ErrUnknownPartition.FastGen(mysql.MySQLErrName[mysql.ErrUnknownPartition], insertPartitionName, e.Table.Meta().Name)
+		}
+	}
+
+	for _, row := range rows {
+		defPos, err := e.Table.LocationPartition(e.ctx, row)
```
Comment (tiancaiamao) [inline executor/insert_common.go]:
```
ok, partitionedTable := e.Table.(PartitionedTable);
if !ok {
        return error...
}
partition, err := partitionedTable.LocationPartition(e.ctx, row) 
```

**Comment by tiancaiamao** [comment]
Close this one as it's already implemented.

**Pull request closed without merging**